File: src/page/awt/BattleBoxDef.java
package page.awt;

import java.awt.Canvas;
import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;

import main.Timer;
import page.battle.BBCtrl;
import page.battle.BattleBox;
import util.basis.BattleField;
import util.basis.SBCtrl;
import util.system.fake.FakeGraphics;
import util.system.fake.awt.FG2D;

class BattleBoxDef extends Canvas implements BattleBox {

	private static final long serialVersionUID = 1L;

	public final BBPainter bbp;

	protected BufferedImage prev;

	protected BattleBoxDef(OuterBox bip, BattleField bas, int p) {
		bbp = p == 0 ? new BBPainter(bip, bas, this) : new BBCtrl(bip, (SBCtrl) bas, this);
		setIgnoreRepaint(true);
	}

	@Override
	public BBPainter getPainter() {
		return bbp;
	}

	@Override
	public void paint() {
		paint(getGraphics());
	}

	@Override
	public void paint(Graphics g) {
		synchronized (bbp) {
			int w = getWidth();
			int h = getHeight();
			if (w * h == 0)
				return;
			if (bbp.pt < bbp.bf.sb.time || prev == null) {
				bbp.pt = bbp.bf.sb.time;
				prev = getImage();
			}
			if (prev == null)
				return;
			g.drawImage(prev, 0, 0, null);
			g.setColor(Color.ORANGE);
			g.drawString("Time cost: " + Timer.inter + "%, " + bbp.pt, 20, 20);
			g.dispose();
		}
	}

	@Override
	public void reset() {
		prev = null;
	}

	protected BufferedImage getImage() {
		int w = getWidth();
		int h = getHeight();
		BufferedImage img = (BufferedImage) createImage(w, h);
		if (img == null)
			return null;
		Graphics2D gra = (Graphics2D) img.getGraphics();
		FakeGraphics g = new FG2D(gra);
		bbp.draw(g);
		gra.dispose();
		return img;
	}

}

File: src/page/awt/BattleBoxDef.kt
package page.awt

import common.battle.BattleField
import common.battle.SBCtrl
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.system.fake.FakeGraphics
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import main.Timer
import page.JL
import page.anim.AnimBox
import page.battle.BBCtrl
import page.battle.BattleBox
import page.battle.BattleBox.BBPainter
import page.battle.BattleBox.OuterBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import utilpc.awt.FG2D
import java.awt.Canvas
import java.awt.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.image.BufferedImage

internal open class BattleBoxDef(bip: OuterBox?, bas: BattleField?, p: Int) : Canvas(), BattleBox {
    val bbp: BBPainter
    protected var prev: BufferedImage? = null
    val painter: BBPainter
        get() = bbp

    override fun paint() {
        paint(graphics)
    }

    override fun paint(g: Graphics) {
        synchronized(bbp) {
            val w = width
            val h = height
            if (w * h == 0) return
            if (bbp.pt < bbp.bf.sb.time || prev == null) {
                bbp.pt = bbp.bf.sb.time
                prev = image
            }
            if (prev == null) return
            g.drawImage(prev, 0, 0, null)
            g.color = Color.ORANGE
            g.drawString("Time cost: " + Timer.Companion.inter + "%, " + bbp.pt, 20, 20)
            g.dispose()
        }
    }

    override fun reset() {
        prev = null
    }

    protected open val image: BufferedImage?
        protected get() {
            val w = width
            val h = height
            val img: BufferedImage = createImage(w, h) as BufferedImage ?: return null
            val gra: Graphics2D = img.getGraphics() as Graphics2D
            val g: FakeGraphics = FG2D(gra)
            bbp.draw(g)
            gra.dispose()
            return img
        }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        bbp = if (p == 0) BBPainter(bip, bas, this) else BBCtrl(bip, bas as SBCtrl?, this)
        ignoreRepaint = true
    }
}
